userspace/libsinsp/filter/parser.cpp

```diff
@@ -0,0 +1,572 @@
+	}
+	throw sinsp_exception("expected a number value");
+}
+
+ast::value_expr* parser::parse_str_value()
+{
+	depth_push();
+	lex_blank();
+	if (lex_quoted_str() || lex_bare_str())
+	{
+		lex_blank();
+		depth_pop();
+		return new ast::value_expr(m_last_token);
+	}
+	throw sinsp_exception("expected a string value");
+}
+
+ast::expr* parser::parse_list_value()
+{
+	depth_push();
+	lex_blank();
+	if (lex_helper_str("("))
+	{
+		bool should_be_empty = false;
+		ast::value_expr* child = NULL;
+		vector<string> values;
+
+		lex_blank();
+		try
+		{
+			child = parse_str_value();
+		}
+		catch(const sinsp_exception& e)
+		{
+			depth_pop();
+			should_be_empty = true;
+		}
+		
+		if (!should_be_empty)
+		{
+			values.push_back(child->value);
+			delete child;
+			lex_blank();
+			while (lex_helper_str(","))
+			{
+				lex_blank();
+				child = parse_str_value();
+				values.push_back(child->value);
+				delete child;
+				lex_blank();
+			}
+		}
+
+		if (!lex_helper_str(")"))
+		{
+			throw sinsp_exception("expected a ')' token");
+		}
+		lex_blank();
+		depth_pop();
+		return new ast::list_expr(values);
+	}
+
+	if (lex_identifier())
+	{
+		lex_blank();
+		depth_pop();
+		return new ast::value_expr(m_last_token);
+	}
+
+	throw sinsp_exception("expected a list or an identifier");
+}
+
+// note: lex_blank is the only lex method that does not update m_last_token.
+bool parser::lex_blank()
+{
+	bool found = false;
+	while(*cursor() == ' ' || *cursor() == '\t' || *cursor() == '\b'
+			|| *cursor() == '\r' || *cursor() == '\n')
+	{
+		found = true;
+		m_pos.col++;
+		if (*cursor() == '\r' || *cursor() == '\n')
+		{
+			m_pos.col = 1;
+			m_pos.line++;
+		}
+		m_pos.idx++;
+	}
+	return found;
+}
+
+inline bool parser::lex_identifier()
+{
+	return lex_helper_rgx("[a-zA-Z]+[a-zA-Z0-9_]*");
+}
+
+inline bool parser::lex_field_name()
+{
+	return lex_helper_rgx("[a-zA-Z]+[a-zA-Z0-9_]*(\\.[a-zA-Z]+[a-zA-Z0-9_]*)+");
+}
+
+inline bool parser::lex_field_arg_bare_str()
+{
+	return lex_helper_rgx("[^ \\b\\t\\n\\r\\[\\]\"']+");
+}
+
+inline bool parser::lex_hex_num()
+{
+	return lex_helper_rgx("0[xX][0-9a-zA-Z]+");
+}
+
+inline bool parser::lex_num()
+{
+	return lex_helper_rgx("[+\\-]?[0-9]+[\\.]?[0-9]*([eE][+\\-][0-9]+)?");
+}
+
+inline bool parser::lex_quoted_str()
+{
+	if(lex_helper_rgx("\"(?:\\\\\"|.)*?\"|'(?:\\\\'|.)*?'"))
+	{
+		m_last_token = escape_str(m_last_token);
+		return true;
+	}
+	return false;
+}
+
+inline bool parser::lex_bare_str()
+{
+	return lex_helper_rgx("[^ \\b\\t\\n\\r\\(\\),=\"']+");
+}
+
+inline bool parser::lex_unary_op()
+{
+	return lex_helper_str_list(unary_ops);
+}
+
+inline bool parser::lex_num_op()
+{
+	return lex_helper_str_list(binary_num_ops);
+}
+
+inline bool parser::lex_str_op()
+{
+	return lex_helper_str_list(binary_str_ops);
+}
+
+inline bool parser::lex_list_op()
+{
+	return lex_helper_str_list(binary_list_ops);
+}
+
+bool parser::lex_helper_rgx(string rgx)
+{
+	cmatch match;
+	auto r = regex("^(" + rgx + ")");
```

Comment:
I remember there being problems with using regex and gcc 4.8 (https://stackoverflow.com/questions/12530406/is-gcc-4-8-or-earlier-buggy-about-regular-expressions). Are we sure that mandating gcc >= 4.9 will be okay?

Reply:
That's a good point. I added an `#ifdef` fallback to use POSIX regex instead of the C++ ones. This should ensure compatibility with older compilers.

Reply:
great thanks!